File: Java/0001-1000/001-100/31-40/35.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        int mid, begin = 0, end = nums.length;
        while (true) {
            if (target < nums[0])
                return 0;
            if (target > nums[nums.length - 1])
                return nums.length;
            mid = (begin + end) / 2;
            if (nums[mid] == target)
                return mid;
            else if (nums[mid] > target) {
                if (mid > 0 && nums[mid - 1] < target)
                    return mid;
                end = mid - 1;
            } else {
                if (mid < nums.length - 1 && nums[mid + 1] > target)
                    return mid + 1;
                begin = mid + 1;
            }
        }
    }
}

File: Python/0001-1000/001-100/31-40/35.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        b = 0
        e = len(nums) - 1
        if target < nums[0]:
            return 0
        if target > nums[len(nums)-1]:
            return len(nums)
        while b <= e:
            mid = (e+b)//2
            if nums[mid] == target:
                return mid
            if nums[mid] > target:
                if nums[mid-1] < target:
                    return mid
                e = mid - 1
            else:
                if nums[mid+1] > target:
                    return mid + 1
                b = mid + 1
        return mid
